### ISSUE TITLE
Meet a dropout missing argument problem when speed up the model and update mask for  my own model 

### ISSUE THREAD

**Issue by sungh66** [microsoft/nni#4297]
**Describe the issue**:
My model is composed of a resnet50 as a base module , F.avg_pool2d and dropout  layer.When speed up the model and update mask for  my  model, i met a dropout layer problem.
my model forward is:
def forward(self, x):
        x = self.base(x)
        x = F.avg_pool2d(x, [7,7])
        x = x.view(x.size(0), -1)
        x = F.dropout(x, p=0.5, training=True)
        x = self.classifier(x)
        return x
I have given the argument 'p=0.5' and 'training=True' ,why TypeError: dropout() missing 2 required positional argument: "p", "train"?

**Environment**:
- NNI version:2.5
- Training service (local|remote|pai|aml|etc):local
- Client OS:ubuntu
- Server OS (for remote mode only):
- Python version:3.7
- PyTorch/TensorFlow version:pytorch
- Is conda/virtualenv/venv used?:conda
- Is running in Docker?:no

**Configuration**:
 - Experiment config (remember to remove secrets!):
 
 - Search space:

**Log message**:
- Traceback (most recent call last):
  File "train_deepmar_resnet50.py", line 670, in <module>
    main(args)
  File "train_deepmar_resnet50.py", line 455, in main
    m_speedup.speedup_model()
  File "/home/xx/anaconda3/lib/python3.7/site-packages/nni/compression/pytorch/speedup/compressor.py", line 507, in speedup_model
    self.infer_modules_masks()
  File "/home/xx/anaconda3/lib/python3.7/site-packages/nni/compression/pytorch/speedup/compressor.py", line 353, in infer_modules_masks
    self.update_direct_sparsity(curnode)
  File "/home/xx/anaconda3/lib/python3.7/site-packages/nni/compression/pytorch/speedup/compressor.py", line 215, in update_direct_sparsity
    func, dummy_input, in_masks, in_constants=in_constants, batch_dim=self.batch_dim)
  File "/home/xx/anaconda3/lib/python3.7/site-packages/nni/compression/pytorch/speedup/infer_mask.py", line 80, in __init__
    self.output = self.module(*dummy_input)
TypeError: dropout() missing 2 required positional argument: "p", "train"
 - nnimanager.log:
 - dispatcher.log:
 - nnictl stdout and stderr:
 

**How to reproduce it?**:

**Comment by J-shang** [comment]
Hi @sungh66 , please try `torch.nn.Dropout2d`. During speedup, it will directly call the function in this way `self.output = self.module(*dummy_input)`, in your case, it will run like this `self.output = F.dropout(*dummy_input)`. So if you have default values in `F.xxx`, please use its `nn.xxx` version.

**Comment by sungh66** [comment]
nn.dropout or nn.dropout2d？They seem different.

**Comment by sungh66** [comment]
nn.Dropout or nn.Dropout2d?They seem different.
Besides, i still met the problem forword missing argument'input', my forward code is following:
```
def forward(self, x):
        x = self.base(x)
        x = F.avg_pool2d(x, [7,7])
        x = x.view(x.size(0), -1)
        m = nn.Dropout2d(p=0.5)
        x = m(x)
        x = self.classifier(x)
      return x
```
The  error log message is following:

>   File "/home/xx/anaconda3/lib/python3.7/site-packages/nni/compression/pytorch/speedup/infer_mask.py", line 80, in __init__
>     self.output = self.module(*dummy_input)
>   File "/home/xx/anaconda3/lib/python3.7/site-packages/torch/nn/modules/module.py", line 727, in _call_impl
>     result = self.forward(*input, **kwargs)
> TypeError: forward() missing 1 required positional argument: 'input'

**Comment by sungh66** [comment]
Hi@J-shang, thanks for your help for my question!!

**Comment by zengjie617789** [comment]
i meet the same error that “TypeError: forward() missing 1 required positional argument: 'input'. I debug this step by step and found the dummy_input is empty when it run at certain modules.
the error show below:
`[2021-11-25 11:18:53] INFO (nni.compression.pytorch.speedup.compressor/MainThread) Update mask for model.base.base_layer.0
[2021-11-25 11:18:53] INFO (nni.compression.pytorch.speedup.compressor/MainThread) Update mask for model.dla_up.ida_0.proj_1.actf.0`

**Comment by J-shang** [comment]
Hi @zengjie617789 , could you show your forward function, and which module do you speed up?

**Comment by zengjie617789** [comment]
@J-shang  hi, Thank you for responsing instantly. I am using the fairmot model to speed up. 
Here is the model structure below:
```
 model = DLASeg('dla{}'.format(num_layers), heads,
                 pretrained=True,
                 down_ratio=down_ratio,
                 final_kernel=1,
                 last_level=5,
                 head_conv=head_conv)
```
the part of foward function:
```
x = self.base(x) 
 x = self.dla_up(x)
```

the base is seemed normal, but the dla_up caused the error:
```
def __init__(self, chi, cho):
        super(DeformConv, self).__init__()
        self.actf = nn.Sequential(
            nn.BatchNorm2d(cho, momentum=BN_MOMENTUM),
            nn.ReLU(inplace=True)
        )
        self.conv = DCN(chi, cho, kernel_size=(3,3), stride=1, padding=1, dilation=1, deformable_groups=1)

    def forward(self, x):
        x = self.conv(x)
        x = self.actf(x)
        return x
```

I found the layer in the meassage  of `Update mask for dla_up.ida_0.proj_1.actf.0` in not in the INFO (FixMaskConflict/MainThread)

**Comment by J-shang** [comment]
From the `forward` function, seems good. Could you find in which place the `dummy_input` become empty, and if the pre-layer output is empty?

Could you give a full error message, such as 

>Traceback (most recent call last):
File "train_deepmar_resnet50.py", line 670, in
main(args)
File "train_deepmar_resnet50.py", line 455, in main
m_speedup.speedup_model()
File "/home/xx/anaconda3/lib/python3.7/site-packages/nni/compression/pytorch/speedup/compressor.py", line 507, in speedup_model
self.infer_modules_masks()
File "/home/xx/anaconda3/lib/python3.7/site-packages/nni/compression/pytorch/speedup/compressor.py", line 353, in infer_modules_masks
self.update_direct_sparsity(curnode)
File "/home/xx/anaconda3/lib/python3.7/site-packages/nni/compression/pytorch/speedup/compressor.py", line 215, in update_direct_sparsity
func, dummy_input, in_masks, in_constants=in_constants, batch_dim=self.batch_dim)
File "/home/xx/anaconda3/lib/python3.7/site-packages/nni/compression/pytorch/speedup/infer_mask.py", line 80, in init
self.output = self.module(*dummy_input)
TypeError: dropout() missing 2 required positional argument: "p", "train"

**Comment by zengjie617789** [comment]
@J-shang Here are the full error message:
```
[2021-11-25 15:33:54] INFO (nni.compression.pytorch.speedup.compressor/MainThread) infer module masks...
[2021-11-25 15:33:54] INFO (nni.compression.pytorch.speedup.compressor/MainThread) Update mask for model.base.base_layer.0
[2021-11-25 15:33:54] INFO (nni.compression.pytorch.speedup.compressor/MainThread) Update mask for model.dla_up.ida_0.proj_1.bn
Traceback (most recent call last):
  File "/home/awifi/zzzj/projects/FairMOT/src/model_pruning.py", line 116, in <module>
    main()
  File "/home/awifi/zzzj/projects/FairMOT/src/model_pruning.py", line 102, in main
    speedup_model()
  File "/home/awifi/zzzj/projects/FairMOT/src/model_pruning.py", line 93, in speedup_model
    m_speedup.speedup_model()
  File "/home/awifi/zzzj/projects/nni/nni/compression/pytorch/speedup/compressor.py", line 507, in speedup_model
    self.infer_modules_masks()
  File "/home/awifi/zzzj/projects/nni/nni/compression/pytorch/speedup/compressor.py", line 353, in infer_modules_masks
    self.update_direct_sparsity(curnode)
  File "/home/awifi/zzzj/projects/nni/nni/compression/pytorch/speedup/compressor.py", line 223, in update_direct_sparsity
    state_dict=copy.deepcopy(module.state_dict()), batch_dim=self.batch_dim)
  File "/home/awifi/zzzj/projects/nni/nni/compression/pytorch/speedup/infer_mask.py", line 80, in __init__
    self.output = self.module(*dummy_input)
  File "/usr/local/miniconda3/envs/yolov5_crowdhuman/lib/python3.6/site-packages/torch/nn/modules/module.py", line 1051, in _call_impl
    return forward_call(*input, **kwargs)
TypeError: forward() missing 1 required positional argument: 'input'
```

**Comment by zengjie617789** [comment]
and the debug results:
![image](https://user-images.githubusercontent.com/25050291/143400400-8e47e48f-37a2-4ac8-af06-e55b310f84be.png)

**Comment by zengjie617789** [comment]
When i changed the pruner algorithum L1Filter to LevelPruner, the first error missed, but occur a new one :
```
  File "/home/awifi/zzzj/projects/nni/nni/compression/pytorch/speedup/compressor.py", line 504, in speedup_model
    fix_mask_conflict(self.masks, self.bound_model, self.dummy_input)
  File "/home/awifi/zzzj/projects/nni/nni/compression/pytorch/utils/mask_conflict.py", line 54, in fix_mask_conflict
    masks = fix_channel_mask.fix_mask()
  File "/home/awifi/zzzj/projects/nni/nni/compression/pytorch/utils/mask_conflict.py", line 304, in fix_mask
    assert self.conv_prune_dim == 0
```

**Comment by J-shang** [comment]
LevelPruner is a fine-grained pruner, speedup may not benefit it.
Will this place `model.dla_up.ida_0.proj_1.bn` has inputs when you inference?
What's the structure of `ida_0`?

**Comment by zengjie617789** [comment]
@J-shang 
the structure of ida is below:
```
class IDAUp(nn.Module):

    def __init__(self, o, channels, up_f):
        super(IDAUp, self).__init__()
        for i in range(1, len(channels)):
            c = channels[i]
            f = int(up_f[i])  
            proj = DeformConv(c, o)
            node = DeformConv(o, o)
     
            up = nn.ConvTranspose2d(o, o, f * 2, stride=f, 
                                    padding=f // 2, output_padding=0,
                                    groups=o, bias=False)
            fill_up_weights(up)

            setattr(self, 'proj_' + str(i), proj)
            setattr(self, 'up_' + str(i), up)
            setattr(self, 'node_' + str(i), node)
                 
        
    def forward(self, layers, startp, endp):
        for i in range(startp + 1, endp):
            upsample = getattr(self, 'up_' + str(i - startp))
            project = getattr(self, 'proj_' + str(i - startp))
            layers[i] = upsample(project(layers[i]))
            node = getattr(self, 'node_' + str(i - startp))
            layers[i] = node(layers[i] + layers[i - 1])
```

**Comment by zengjie617789** [comment]
I wonder if  the output of DCN is empty that caused the problem?
```
class DeformConv(nn.Module):
    def __init__(self, chi, cho):
        super(DeformConv, self).__init__()
        self.actf = nn.Sequential(
            nn.BatchNorm2d(cho, momentum=BN_MOMENTUM),
            nn.ReLU(inplace=True)
        )
        self.conv = DCN(chi, cho, kernel_size=(3,3), stride=1, padding=1, dilation=1, deformable_groups=1)

    def forward(self, x):
        x = self.conv(x)
        x = self.actf(x)

        return x
```

**Comment by J-shang** [comment]
yes, if the output of DCN is empty, then speedup won't get the input of `actf`.
but I don't know why it can inference successful if the output of DCN is empty.

**Comment by zengjie617789** [comment]
@J-shang thank you for your help anyway. I debug this code step by step, i found that the node of `model.dla_up.ida_0.proj_1.actf.0` need the node of `input.223` which is not exist. I record the message of related node which is that `input.221', 'out.25', 'input.225', 'input.227',`.  I think it is about the output of DCN layer . Moreover, I changed the way to prune the base part of the model which exclude the DCN layer. But I got the error about out of CUDA memory.
Here is the code below:
```
    input_model.eval()
    m_speedup = ModelSpeedup(input_model, dummy_input, masks_file)
    m_speedup.speedup_model()
```
